File: Cpp/CXR_classify/confighandler.cpp
#include "confighandler.h"

namespace fs = std::filesystem;

ConfigHandler::ConfigHandler(std::string configFilename)
{
    ConfigHandler::configFilename = configFilename;
    readConfigFile();
    prepConfigIni();
}

ConfigHandler::~ConfigHandler()
{
    boost::property_tree::ini_parser::write_ini(configFilename, configFile);
}

boost::property_tree::ptree ConfigHandler::getSection(std::string sectionName)
{
    return configFile.get_child(sectionName);
}

std::string ConfigHandler::getSetting(std::string sectionName, std::string settingName)
{
    boost::property_tree::ptree section = ConfigHandler::getSection(sectionName);
    return section.get<std::string>(settingName);
}

void ConfigHandler::setSetting(std::string sectionName, std::string settingName, std::string value)
{
    configFile.put(sectionName + "." + settingName, value);
}

void ConfigHandler::prepConfigIni()
{
    setUrl(sourceUrl.at(getDatasetType()));
    setParentFolder();
    setCsvName();
    setColumnsInfoName();
}

void ConfigHandler::readConfigFile()
{
    if (fs::exists(configFilename)) {
        boost::property_tree::ini_parser::read_ini(configFilename, configFile);
    }
}

void ConfigHandler::setUrl(std::string url)
{
    setSetting("misc", "url", url);
}

void ConfigHandler::setParentFolder()
{
    setSetting("misc", "parent_folder", fs::current_path());
}

void ConfigHandler::setCsvName()
{
    setSetting("misc", "csv_filename", "image_labels.csv");
}

void ConfigHandler::setColumnsInfoName()
{
    setSetting("misc", "columns_info_name", "columns_info.json");
}

boost::property_tree::ptree ConfigHandler::getDbInfo()
{
    return getSection("postgresql");
}

std::string ConfigHandler::getTableName(std::string table)
{
    return getSetting("table_names", table);
}

std::string ConfigHandler::getUrl()
{
    return getSetting("misc", "url");
}

std::string ConfigHandler::getTgzFilename()
{
    return getUrl().substr(getUrl().find_last_of("/") + 1);
}

std::string ConfigHandler::getTgzFilePath()
{
    return prependParentPath(fs::path(getTgzFilename()));
}

std::string ConfigHandler::getDatasetName()
{
    return getTgzFilename().substr(0, getTgzFilename().find_last_of("."));
}

std::string ConfigHandler::getUnpackFolderPath()
{
    return prependParentPath(fs::path(getDatasetName()));
}

std::string ConfigHandler::getColumnsInfoName()
{
    return getSetting("misc", "columns_info_name");
}

std::string ConfigHandler::getColumnsInfoPath()
{
    return prependParentPath(fs::path(getCsvName()));
}

std::string ConfigHandler::getCsvName()
{
    return getSetting("misc", "csv_filename");
}

std::string ConfigHandler::getCsvPath()
{
    return prependParentPath(getCsvName());
}

std::string ConfigHandler::getDatasetType()
{
    return getSetting("dataset_info", "dataset");
}

std::string ConfigHandler::getParentFolder()
{
    return getSetting("misc", "parent_folder");
}

std::string ConfigHandler::getLogLevel()
{
    return getSetting("logging", "level");
}

std::string ConfigHandler::getConfigFilename()
{
    return configFilename;
}

std::string ConfigHandler::getConfigFilePath()
{
    return prependParentPath(getConfigFilename());
}

std::string ConfigHandler::prependParentPath(std::string fsItem)
{
    return (fs::path(getParentFolder()) / fs::path(fsItem));
}

File: Go/CxrClassify/configHandler/configHandler.go
package configHandler

import (
	"expectedSizes"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/ini.v1"
)

type ConfigHandler struct {
	configFilename string
	configFile     ini.File
}

func NewConfigHandler(configFilename string) *ConfigHandler {
	c := new(ConfigHandler)
	c.configFilename = configFilename
	c.readConfigFile()
	c.prepConfigIni()
	return c
}

func (c ConfigHandler) Close() {
	c.configFile.SaveTo("test.ini") // change to c.configFilename
}

func (c ConfigHandler) getSection(sectionName string) ini.Section {
	return *c.configFile.Section(sectionName)
}

func (c ConfigHandler) getSetting(sectionName, settingName string) string {
	section := c.getSection(sectionName)
	return section.Key(settingName).String()
}

func (c ConfigHandler) setSetting(sectionName, settingName, value string) {
	c.configFile.Section(sectionName).Key(settingName).SetValue(value)
}

func (c ConfigHandler) prepConfigIni() {
	c.setUrl(expectedSizes.SourceUrl[c.GetDatasetType()])
	c.setParentFolder()
	c.setCsvName()
	c.setColumnsInfoName()
}

func (c *ConfigHandler) readConfigFile() {
	if _, err := os.Stat(c.configFilename); err == nil {
		// path/to/whatever exists
		cfg, _ := ini.Load(c.configFilename)
		c.configFile = *cfg
	}
}

func (c ConfigHandler) setUrl(url string) {
	c.setSetting("misc", "url", url)
}

func (c ConfigHandler) setParentFolder() {
	path, _ := os.Getwd()
	c.setSetting("misc", "parent_folder", path)
}

func (c ConfigHandler) setCsvName() {
	c.setSetting("misc", "csv_filename", "image_labels.csv")
}

func (c ConfigHandler) setColumnsInfoName() {
	c.setSetting("misc", "columns_info_name", "columns_info.json")
}

func (c ConfigHandler) GetDbInfo() ini.Section {
	return c.getSection("postgresql")
}

func (c ConfigHandler) GetTableName(table string) string {
	return c.getSetting("table_names", table)
}

func (c ConfigHandler) GetUrl() string {
	return c.getSetting("misc", "url")
}

func (c ConfigHandler) GetTgzFilename() string {
	url := c.GetUrl()
	basename := filepath.Base(url)
	return basename
}

func (c ConfigHandler) GetTgzFilePath() string {
	return c.PrependParentPath(c.GetTgzFilename())
}

func (c ConfigHandler) GetDatasetName() string {
	basename := c.GetTgzFilename()
	nameWithoutExt := strings.TrimSuffix(basename, filepath.Ext(basename))
	return nameWithoutExt
}

func (c ConfigHandler) GetUnpackFolderPath() string {
	return c.PrependParentPath(c.GetDatasetName())
}

func (c ConfigHandler) GetColumnsInfoName() string {
	return c.getSetting("misc", "columns_info_name")
}

func (c ConfigHandler) GetColumnsInfoPath() string {
	return c.PrependParentPath(c.GetColumnsInfoName())
}

func (c ConfigHandler) GetCsvName() string {
	return c.getSetting("misc", "csv_filename")
}

func (c ConfigHandler) GetCsvPath() string {
	return c.PrependParentPath(c.GetCsvName())
}

func (c ConfigHandler) GetDatasetType() string {
	return c.getSetting("dataset_info", "dataset")
}

func (c ConfigHandler) GetParentFolder() string {
	return c.getSetting("misc", "parent_folder")
}

func (c ConfigHandler) GetLogLevel() string {
	return c.getSetting("logging", "level")
}

func (c ConfigHandler) GetConfigFilename() string {
	return c.configFilename
}

func (c ConfigHandler) GetConfigFilePath() string {
	return c.PrependParentPath(c.GetConfigFilename())
}

func (c ConfigHandler) PrependParentPath(fsItem string) string {
	strArr := []string{c.GetParentFolder(), fsItem}
	return strings.Join(strArr, "\\")
}
